File: cpp/Medium/238_product_of_array_except_self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int size = nums.size();
        int left = 1;
        vector<int>right(size,1);
        
        for( int i = size-2; i >= 0; i--){
            right[i] = right[i+1] * nums[i+1]; 
        }
        for( int i = 1; i < size; i++){
            left = left * nums[i-1];
            right[i] = left * right[i];
        }
        return right;
    }
};

File: python/medium/238_product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        answer = [1] * len(nums)
        pre = 1
        post = 1
        for i in range(len(nums)):
            answer[i] = pre
            pre *= nums[i]
        for i in range(len(nums) -1, -1, -1):
            answer[i] *= post
            post *= nums[i]
        return answer

